activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb

```diff
@@ -528,12 +524,27 @@ def combine_multi_statements(total_sql)
         end
 
         # Returns an ActiveRecord::Result instance.
-        def select(sql, name = nil, binds = [])
-          exec_query(sql, name, binds, prepare: false)
-        end
+        def select(sql, name = nil, binds = [], prepare: false, async: false)
+          if async
+            if current_transaction.joinable?
+              raise AsynchronousQueryInsideTransaction, "Asynchronous queries are not allowed inside transactions"
+            end
 
-        def select_prepared(sql, name = nil, binds = [])
-          exec_query(sql, name, binds, prepare: true)
+            future_result = async.new(
+              pool,
+              sql,
+              name,
+              binds,
+              prepare: prepare,
+            )
+            if supports_concurrent_connections? && current_transaction.closed? && ActiveRecord::Base.asynchronous_queries_session
```

Comment:
It feels wrong to reach to `ActiveRecord::Base` here. But I'm unsure how to get rid of this.